wdl-grammar/src/v1/lint/version_declaration_placement.rs

```diff
@@ -0,0 +1,214 @@
+
+use std::collections::VecDeque;
+
+use nonempty::NonEmpty;
+use pest::iterators::Pair;
+use wdl_core::concern::code;
+use wdl_core::concern::lint;
+use wdl_core::concern::lint::Group;
+use wdl_core::concern::lint::Rule;
+use wdl_core::concern::Code;
+use wdl_core::file::Location;
+use wdl_core::Version;
+
+use crate::v1;
+
+/// Detects an improperly placed version declaration.
+#[derive(Debug)]
+pub struct VersionDeclarationPlacement;
+
+impl<'a> VersionDeclarationPlacement {
+    /// Generates a validation error for an improperly placed version
+    /// declaration.
+    fn misplaced_version(&self, location: Location) -> lint::Warning
+    where
+        Self: Rule<&'a Pair<'a, v1::Rule>>,
+    {
+        // SAFETY: this error is written so that it will always unwrap.
+        lint::warning::Builder::default()
+            .code(self.code())
+            .level(lint::Level::Medium)
+            .group(lint::Group::Spacing)
+            .subject("Improperly placed version declaration")
+            .body(
+                "The version declaration must be the first line of a WDL document or one blank \
+                 line after header comments.",
+            )
+            .push_location(location)
+            .fix(
+                "Move the version declaration to the first line of the WDL document or one blank \
+                 line after header comments.",
+            )
+            .try_build()
+            .unwrap()
+    }
+}
+
+impl<'a> Rule<&Pair<'a, v1::Rule>> for VersionDeclarationPlacement {
+    fn code(&self) -> Code {
+        // SAFETY: this manually crafted to unwrap successfully every time.
+        Code::try_new(code::Kind::Error, Version::V1, 9).unwrap()
+    }
+
+    fn group(&self) -> Group {
+        Group::Spacing
+    }
+
+    fn check(&self, tree: &Pair<'_, v1::Rule>) -> lint::Result {
+        let mut warnings = VecDeque::new();
+
+        // Optionally consume comment nodes, if found, allow exactly one empty line
+        // between the comment and the version declaration.
+        let mut comment = 0;
+        let mut newline = 0;
+        let mut anything_else = 0;
+
+        // This will never get used. Validation rules require a version statement.
+        let mut location: Location = Location::Unplaced;
+
+        for node in tree.clone().into_inner() {
+            match node.as_rule() {
+                v1::Rule::COMMENT => {
+                    comment += 1;
+                }
+                v1::Rule::WHITESPACE => {
+                    if node.as_str() == "\n" {
+                        newline += 1;
+                    } else {
+                        anything_else += 1;
+                    }
+                }
+                v1::Rule::version => {
+                    location = Location::try_from(node.as_span()).map_err(lint::Error::Location)?;
+                    break;
+                }
+                _ => {
+                    anything_else += 1;
```

Comment:
Should this be `unreachable!()`? When could something other than `COMMENT` or `WHITESPACE` be found before `version`?

Reply:
It looks like the lints are being run as part of some of the other tests, e.g. https://github.com/stjude-rust-labs/wdl/blob/35731eca283d4aba4b7349910a19d579df837ac1/wdl-ast/src/v1/document/declaration/bound.rs#L186. Since those are testing individual nodes, there is no `version` declaration. I'm not sure if the lints should be running, I would have thought not. Maybe @claymcleod can weigh in.

Reply:
I think, for individual nodes, the lints can be disabled.

Reply:
I'm not clear how they're being run (or how to disable, if that's what should happen). It's when I run `cargo test --all-features`, but it produces a long list of failures, most of which are `it_parses_from_a_supported_node_type` type tests. 

```
    v1::document::declaration::bound::tests::it_parses_from_a_supported_node_type
    v1::document::declaration::unbound::tests::it_parses_from_a_supported_node_type
    v1::document::expression::array::tests::it_parses_from_a_supported_node_type
    v1::document::expression::map::tests::it_parses_from_a_supported_node_type
    v1::document::expression::pair::tests::it_parses_from_a_supported_node_type
    v1::document::expression::r#if::tests::it_parses_from_a_supported_node_type
    v1::document::expression::tests::ensure_number_works_correctly
    v1::document::expression::tests::it_correctly_parses_floats
    v1::document::expression::tests::it_ignores_comments_and_whitespace
    v1::document::identifier::qualified::tests::it_parses_from_a_supported_node_type
    v1::document::identifier::tests::it_parses_from_a_supported_node_type
    v1::document::import::tests::it_parses_a_complicated_import_correctly
    v1::document::output::builder::tests::it_parses_from_a_supported_node_type
    v1::document::private_declarations::tests::it_parses_from_a_supported_node_type
    v1::document::r#struct::tests::it_parses_from_a_supported_node_type
    v1::document::task::runtime::value::tests::it_correctly_parses_floats
    v1::document::task::runtime::value::tests::it_correctly_parses_integers
    v1::document::tests::it_parses_from_a_supported_node_type
    v1::document::workflow::execution::statement::call::body::tests::it_parses_from_a_supported_node_type
    v1::document::workflow::execution::statement::call::tests::it_parses_from_a_supported_node_type
    v1::document::workflow::execution::statement::conditional::tests::it_parses_from_a_supported_node_type
    v1::document::workflow::execution::statement::scatter::tests::it_parses_from_a_supported_node_type
```

Reply:
Ah, I see. It's here in `parse_rule`. https://github.com/stjude-rust-labs/wdl/blob/35731eca283d4aba4b7349910a19d579df837ac1/wdl-grammar/src/v1.rs#L177-L179

Reply:
I don't see a way to disable running the lint rules in that function. It does appear that nothing is done with any lint warnings. Was this intentional?